src/emqtt.erl

```diff
@@ -1252,6 +1436,24 @@ delete_inflight_when_full(Packet, State) ->
         false -> {next_state, connected, State1}
     end.
 
+drop_expired(Pendings = #{count := 0}) ->
+    Pendings;
+drop_expired(Pendings) ->
+    drop_expired(Pendings, now_ts()).
+
+drop_expired(Pendings = #{count := 0}, _Now) ->
+    Pendings;
+drop_expired(Pendings = #{requests := Reqs}, Now) ->
+    {value, ?PUB_REQ(_Msg, ExpireAt, Callback)} = queue:peek(Reqs),
+    case Now > ExpireAt of
+        true ->
+            {_Dropped, NPendings} = dequeue_publish_req(Pendings),
+            eval_callback_handler({error, timeout}, Callback),
+            drop_expired(NPendings, Now);
+        false ->
+            Pendings
```

Comment:
If each message has its own `ExpiredAt` and they are not necessarily sorted by that in the queue, maybe we need to keep checking the whole queue?

Reply:
I have refactored inflight as suggested by stone. Now the resend and iterations are in sequence. 😆